runtime/sema/type_tags.go

```diff
@@ -719,6 +720,132 @@ func findSuperTypeFromUpperMask(joinedTypeTag TypeTag, types []Type) Type {
+			// E.g., `(A | C)`, `(A | D)`, `(B | C)`, and `(B | D)`
+			// are all equally valid least common supertypes of (A, B)` and `(C, D)`.
+			//
+			// To get a more consistent behavior here,
+			// we take the least common supertype of all the possible least common supertypes from the previous step,
+			// which luckily here is just the union of the elements of the conjunctions converted to a disjunction.
+			// e.g. the least common supertype of E and F is `(E | F)`
+			// and the least common supertype of `(A, B)` and `(C, D)` is `(A | B | C | D)`
+			union := orderedmap.KeySetUnion(setAccessA.Entitlements, setAccessB.Entitlements)
+			return NewAccessFromEntitlementSet(union, Disjunction)
+
+		case Disjunction:
+			// least common supertype of a non-disjoint conjunction and a disjunction is
+			// just the disjunction. This is because, in this case, the disjunction is
+			// already a supertype of the conjunction
+			// e.g. the least common access of `(E, F)` and `(E | G)` is `(E | G)`
+			if setAccessB.PermitsAccess(setAccessA) {
+				return setAccessB
+			}
+
+			// if the conjunction and disjunction are completely disjoint, their least common supertype
+			// is the union of the disjunction and an arbitrary element chosen from the conjunction.
+			// E.g. `(E | G | H)` and `(F | G | H)` are both valid least common supertypes of `(E, F)` and `(G | H)`
+			//
+			// In order to have a consistent behavior here,
+			// we take the least common supertype of all the possible least common supertypes from the previous step,
+			// which luckily here is just the union of the elements of the disjunction and the conjunction.
+			// E.g. our computed supertype of `(E, F)` and `(G | H)` is `(E | F | G | H)`
+			union := orderedmap.KeySetUnion(setAccessA.Entitlements, setAccessB.Entitlements)
+			return NewAccessFromEntitlementSet(union, Disjunction)
+		}
+
+	case Disjunction:
+		switch setAccessB.SetKind {
+		case Conjunction:
+			// symmetric with the other case where A is a conjunction and B is a disjunction
+			return leastCommonAccess(accessB, accessA)
+		case Disjunction:
+			// least common access of two disjunctions is their union
+			// e.g. the least common supertype of (E | F) and (E | G) is (E | F | G)
+			union := orderedmap.KeySetUnion(setAccessA.Entitlements, setAccessB.Entitlements)
+			return NewAccessFromEntitlementSet(union, Disjunction)
+		}
+	}
+
+	return UnauthorizedAccess
+}
+
+func commonSuperTypeOfReferences(types []Type) Type {
+	var references []*ReferenceType
+
+	// check that all the referenced types are equal
+	// before computing authorization supertype
+	var prevReferenceType *ReferenceType
+	for _, typ := range types {
+		// 'Never' type doesn't affect the supertype.
+		// Hence, ignore them
+		if typ == NeverType {
+			continue
+		}
+
+		referenceType, ok := typ.(*ReferenceType)
+		if !ok {
+			return getSuperTypeOfDerivedTypes(types)
```

Comment:
Can maybe directly use `commonSuperTypeOfHeterogeneousTypes` since we know the two types are not the same 'kind'?